x/twap/listeners_test.go

```diff
@@ -77,58 +78,220 @@ func (s *TestSuite) TestAfterPoolCreatedHook() {
+	twapBeforeEpoch, err := s.twapkeeper.GetRecordAtOrBeforeTime(s.Ctx, poolId, s.Ctx.BlockTime(), denom0, denom1)
 	s.Require().NoError(err)
+	pruneEpochIdentifier := s.App.TwapKeeper.PruneEpochIdentifier(s.Ctx)
 
-	s.EndBlock()
-	s.Commit() // clear transient store
-	// Now on a clean state after a create pool
-	s.Require().Equal(baseTime.Add(time.Second), s.Ctx.BlockTime())
-	s.RunBasicSwap(poolId)
+	// make prune record time pass by, running prune epoch after this should prune old record
+	s.Ctx = s.Ctx.WithBlockTime(baseTime.Add(twap.RecordHistoryKeepPeriod).Add(time.Second))
 
-	// accumulators are default right here
-	expectedLatestTwapUpToAccum, err := types.NewTwapRecord(s.App.GAMMKeeper, s.Ctx, poolId, denom0, denom1)
-	s.Require().NoError(err)
-	// ensure different spot prices
-	s.Require().NotEqual(expectedHistoricalTwap.P0LastSpotPrice, expectedLatestTwapUpToAccum.P0LastSpotPrice)
-	s.Require().NotEqual(expectedHistoricalTwap.P1LastSpotPrice, expectedLatestTwapUpToAccum.P1LastSpotPrice)
+	allEpochs := s.App.EpochsKeeper.AllEpochInfos(s.Ctx)
 
-	s.EndBlock()
+	// iterate through all epoch, ensure that epoch only gets pruned in prune epoch identifier
+	// we reverse iterate here to test epochs that are not prune epoch
+	for i := len(allEpochs) - 1; i >= 0; i-- {
+		s.App.TwapKeeper.EpochHooks().AfterEpochEnd(s.Ctx, allEpochs[i].Identifier, int64(1))
 
-	// check records
-	historicalOldTwap, err := s.twapkeeper.GetRecordAtOrBeforeTime(s.Ctx, poolId, baseTime, denom0, denom1)
-	s.Require().NoError(err)
-	s.Require().Equal(expectedHistoricalTwap, historicalOldTwap)
+		recentTwapRecords, err := s.twapkeeper.GetRecordAtOrBeforeTime(s.Ctx, poolId, baseTime.Add(allEpochs[i].Duration), denom0, denom1)
 
-	latestTwap, err := s.twapkeeper.GetMostRecentRecordStoreRepresentation(s.Ctx, poolId, denom0, denom1)
-	s.Require().NoError(err)
-	s.Require().Equal(latestTwap.P0LastSpotPrice, expectedLatestTwapUpToAccum.P0LastSpotPrice)
-	s.Require().Equal(latestTwap.P1LastSpotPrice, expectedLatestTwapUpToAccum.P1LastSpotPrice)
+		// old record should have been pruned here
+		if allEpochs[i].Identifier == pruneEpochIdentifier {
+			s.Require().Error(err)
+			s.Require().NotEqual(twapBeforeEpoch, recentTwapRecords)
 
-	latestTwap2, err := s.twapkeeper.GetRecordAtOrBeforeTime(s.Ctx, poolId, s.Ctx.BlockTime(), denom0, denom1)
-	s.Require().NoError(err)
-	s.Require().Equal(latestTwap, latestTwap2)
+			// quit test once the record has been pruned
+			return
+		} else { // pruning should not be triggered at first, not pruning epoch
+			s.Require().NoError(err)
+			s.Require().Equal(twapBeforeEpoch, recentTwapRecords)
+		}
+	}
+}
+
+// TestAfterSwap_JoinPool tests hooks for `AfterSwap` and `AfterJoinPool`.
+// The purpose of this test is to test whether we correctly store the state of the
+// pools that has changed with price impact.
+func (s *TestSuite) TestAfterSwap_JoinPool() {
+	two_asset_coins := defaultTwoAssetCoin
+	three_asset_coins := defaultThreeAssetCoins
```

Comment:
can we just use defaultTwoAssetCoin and defaultThreeAssetCoins directly?

Reply:
fixed!